File: go/backtracking/46_permute.go
package backtracking

import "slices"

// permute 全排列
func permute(nums []int) [][]int {

	res := make([][]int, 0)

	path := make([]int, 0, len(nums))

	var backtrack func()
	backtrack = func() {
		if len(path) == len(nums) {
			p := make([]int, 0, len(nums))
			copy(p, path)
			res = append(res, p)
			return
		}

		for i := 0; i < len(nums); i++ {
			if !slices.Contains(path, nums[i]) {
				path = append(path, nums[i])
				backtrack()
				path = path[:len(path)-1]
			}
		}

	}

	backtrack()
	return res
}

File: rust/src/backtracking/_46_permute.rs
/// 全排列
pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut res = vec![];
    let mut path = vec![];

    fn backtrack(nums: &[i32], path: &mut Vec<i32>, res: &mut Vec<Vec<i32>>) {
        if path.len() == nums.len() {
            res.push(path.clone());
            return;
        }

        for item in nums.iter() {
            if !path.contains(item) {
                path.push(*item);
                backtrack(nums, path, res);
                path.pop();
            }
        }
    }

    backtrack(&nums, &mut path, &mut res);
    res
}
